scanpy/metrics/_gearys_c.py

```diff
@@ -0,0 +1,286 @@
+#         sk = slice(x_indptr[k], x_indptr[k + 1])
+#         x_arr[x_indices[sk]] = x_data[sk]
+#         outval = _gearys_c_vec_W(g_data, g_indices, g_indptr, x_arr, W)
+#         out[k] = outval
+#     return out
+
+
+@numba.njit(cache=True, parallel=True)
+def _gearys_c_mtx(g_data, g_indices, g_indptr, X):
+    M, N = X.shape
+    W = g_data.sum()
+    out = np.zeros(M, dtype=np.float_)
+    for k in numba.prange(M):
+        x = X[k, :].astype(np.float_)
+        x_bar = x.mean()
+
+        total = 0.0
+        for i in numba.prange(N):
+            s = slice(g_indptr[i], g_indptr[i + 1])
+            i_indices = g_indices[s]
+            i_data = g_data[s]
+            total += np.sum(i_data * ((x[i] - x[i_indices]) ** 2))
+
+        numer = (N - 1) * total
+        denom = 2 * W * ((x - x_bar) ** 2).sum()
+        C = numer / denom
+
+        out[k] = C
+    return out
+
+
+# Similar to above, simplified version umaps choice of parallel backend breaks:
+# @numba.njit(cache=True, parallel=True)
+# def _gearys_c_mtx(g_data, g_indices, g_indptr, X):
+#     M, N = X.shape
+#     W = g_data.sum()
+#     out = np.zeros(M, dtype=np.float64)
+#     for k in numba.prange(M):
+#         outval = _gearys_c_vec_W(g_data, g_indices, g_indptr, X[k, :], W)
+#         out[k] = outval
+#     return out
+
+
+###############################################################################
+# Interface
+###############################################################################
+
+
+@dispatch(sparse.csr_matrix, sparse.csr_matrix)
+def gearys_c(g, vals) -> np.ndarray:
+    assert g.shape[0] == g.shape[1], "`g` should be a square adjacency matrix"
+    assert g.shape[0] == vals.shape[1]
+    return _gearys_c_mtx_csr(
+        g.data.astype(np.float_, copy=False),
+        g.indices,
+        g.indptr,
+        vals.data.astype(np.float_, copy=False),
+        vals.indices,
+        vals.indptr,
+        vals.shape,
+    )
+
+
+@dispatch(sparse.spmatrix, np.ndarray)  # noqa
+def gearys_c(g, vals):
+    assert g.shape[0] == g.shape[1], "`g` should be a square matrix."
+    if not isinstance(g, sparse.csr_matrix):
+        g = g.tocsr()
+    g_data = g.data.astype(np.float_, copy=False)
+    if vals.ndim == 1:
+        assert g.shape[0] == vals.shape[0]
+        return _gearys_c_vec(g_data, g.indices, g.indptr, vals)
+    elif vals.ndim == 2:
+        assert g.shape[0] == vals.shape[1]
+        return _gearys_c_mtx(g_data, g.indices, g.indptr, vals)
+    else:
+        raise ValueError()
+
+
+@dispatch(sparse.spmatrix, (pd.DataFrame, pd.Series))  # noqa
+def gearys_c(g, vals):
+    return gearys_c(g, vals.values)
+
+
+@dispatch(sparse.spmatrix, sparse.spmatrix)  # noqa
+def gearys_c(g, vals) -> np.ndarray:
+    if not isinstance(g, sparse.csr_matrix):
+        g = g.tocsr()
+    if not isinstance(vals, sparse.csr_matrix):
+        vals = vals.tocsr()
+    return gearys_c(g, vals)
+
+
+# TODO: Document better
+# TODO: Have scanpydoc work with multipledispatch
```

Comment:
Changes to scanpydoc are online fast, do you have something in line?

Reply:
I'm pretty open to suggestions on this.

We could do something similar to R? There they give all the signatures and then have all the arguments for all of them mixed together.

Is it possible to have multiple `Params` sections in sphinx? Otherwise, maybe we find functions which have had multiple dispatch applied, then know to stitch their doc strings together somehow?

I feel like languages that use dispatch more heavily (and not like `object.method`) tend to have more narrative docs with examples of usage.

Reply:
What’s the current output? Do we at least have the docstring and signature of one of them?

Reply:
`?sc.metrics.gearys_c` looks like:

<details>
<summary> Dispatched docstring </summary>

```rst
[0;31mSignature:[0m   [0msc[0m[0;34m.[0m[0mmetrics[0m[0;34m.[0m[0mgearys_c[0m[0;34m([0m[0;34m*[0m[0margs[0m[0;34m,[0m [0;34m**[0m[0mkwargs[0m[0;34m)[0m[0;34m[0m[0;34m[0m[0m
[0;31mType:[0m        Dispatcher
[0;31mString form:[0m <dispatched gearys_c>
[0;31mFile:[0m        /usr/local/lib/python3.7/site-packages/multipledispatch/dispatcher.py
[0;31mDocstring:[0m  
Multiply dispatched method: gearys_c

Inputs: <AnnData>
------------------
Calculate `Geary's C` <https://en.wikipedia.org/wiki/Geary's_C>`_, as used
    by `VISION <https://doi.org/10.1038/s41467-019-12235-0>`_.

    Geary's C is a measure of autocorrelation for some measure on a graph. This
    can be to whether measures are correlated between neighboring cells. Lower
    values indicate greater correlation.

    ..math

        C =
        rac{
            (N - 1)\sum_{i,j} w_{i,j} (x_i - x_j)^2
        }{
            2W \sum_i (x_i - ar{x})^2
        }

    Params
    ------
    adata
    vals
        Values to calculate Geary's C for. If this is two dimensional, should
        be of shape `(n_features, n_cells)`. Otherwise should be of shape
        `(n_cells,)`. This matrix can be selected from elements of the anndata
        object by using key word arguments: `layer`, `obsm`, `obsp`, or
        `use_raw`.
    use_graph
        Key to use for graph in anndata object. If not provided, default
        neighbors connectivities will be used instead.
    layer
        Key for `adata.layers` to choose `vals`.
    obsm
        Key for `adata.obsm` to choose `vals`.
    obsp
        Key for `adata.obsp` to choose `vals`.
    use_raw
        Whether to use `adata.raw.X` for `vals`.

    Returns
    -------
    If vals is two dimensional, returns a 1 dimensional ndarray array. Returns
    a scalar if `vals` is 1d.

Other signatures:
    spmatrix, spmatrix
    csr_matrix, csr_matrix
    spmatrix, ndarray
    spmatrix, DataFrame
    spmatrix, Series
```

</details>

So not that much of a signature. I added a doc-string to one of the other methods and got this:

<details>
<summary> Two doc-strings </summary>

```rst
[0;31mSignature:[0m   [0msc[0m[0;34m.[0m[0mmetrics[0m[0;34m.[0m[0mgearys_c[0m[0;34m([0m[0;34m*[0m[0margs[0m[0;34m,[0m [0;34m**[0m[0mkwargs[0m[0;34m)[0m[0;34m[0m[0;34m[0m[0m
[0;31mType:[0m        Dispatcher
[0;31mString form:[0m <dispatched gearys_c>
[0;31mFile:[0m        /usr/local/lib/python3.7/site-packages/multipledispatch/dispatcher.py
[0;31mDocstring:[0m  
Multiply dispatched method: gearys_c

Inputs: <spmatrix, ndarray>
----------------------------
Params
    ------
    g
        Connectivity graph as a scipy sparse matrix. Should have shape:
        `(n_obs, n_obs)`.
    vals
        Values to calculate Geary's C for. If one dimensional, should have
        shape `(n_obs,)`. If two dimensional (i.e calculating Geary's C for
        multiple variables) should have shape `(n_vars, n_obs)`.

Inputs: <AnnData>
------------------
Calculate `Geary's C` <https://en.wikipedia.org/wiki/Geary's_C>`_, as used
    by `VISION <https://doi.org/10.1038/s41467-019-12235-0>`_.

    Geary's C is a measure of autocorrelation for some measure on a graph. This
    can be to whether measures are correlated between neighboring cells. Lower
    values indicate greater correlation.

    ..math

        C =
        rac{
            (N - 1)\sum_{i,j} w_{i,j} (x_i - x_j)^2
        }{
            2W \sum_i (x_i - ar{x})^2
        }

    Params
    ------
    adata
    vals
        Values to calculate Geary's C for. If this is two dimensional, should
        be of shape `(n_features, n_cells)`. Otherwise should be of shape
        `(n_cells,)`. This matrix can be selected from elements of the anndata
        object by using key word arguments: `layer`, `obsm`, `obsp`, or
        `use_raw`.
    use_graph
        Key to use for graph in anndata object. If not provided, default
        neighbors connectivities will be used instead.
    layer
        Key for `adata.layers` to choose `vals`.
    obsm
        Key for `adata.obsm` to choose `vals`.
    obsp
        Key for `adata.obsp` to choose `vals`.
    use_raw
        Whether to use `adata.raw.X` for `vals`.

    Returns
    -------
    If vals is two dimensional, returns a 1 dimensional ndarray array. Returns
    a scalar if `vals` is 1d.

Other signatures:
    spmatrix, spmatrix
    csr_matrix, csr_matrix
    spmatrix, DataFrame
    spmatrix, Series

```

</details>

Reply:
Scanpydoc throws an error if I try to add this to the docs. The full log is below, but I think these are the relevant parts of the traceback:

```python
  File "/usr/local/lib/python3.7/site-packages/sphinx/events.py", line 103, in emit
    results.append(callback(self.app, *args))
  File "/usr/local/lib/python3.7/site-packages/scanpydoc/elegant_typehints/return_tuple.py", line 46, in process_docstring
    ret_types = get_tuple_annot(get_type_hints(obj).get("return"))
  File "/usr/local/Cellar/python/3.7.5/Frameworks/Python.framework/Versions/3.7/lib/python3.7/typing.py", line 996, in get_type_hints
    'or function.'.format(obj))
TypeError: <dispatched gearys_c> is not a module, class, method, or function.
```

<details>
<summary> Full error </summary>

```python
# Sphinx version: 2.2.1
# Python version: 3.7.5 (CPython)
# Docutils version: 0.15.2 release
# Jinja2 version: 2.10.1
# Last messages:
#   loading pickled environment...
#   done
#   [autosummary] generating autosummary for: api/index.rst, api/plotting.rst, api/scanpy.Neighbors.compute_eigen.rst, api/scanpy.Neighbors.compute_neighbors.rst, api/scanpy.Neighbors.compute_transitions.rst, api/scanpy.Neighbors.connectivities.rst, api/scanpy.Neighbors.distances.rst, api/scanpy.Neighbors.distances_dpt.rst, api/scanpy.Neighbors.eigen_basis.rst, api/scanpy.Neighbors.eigen_values.rst, ..., external/scanpy.external.tl.cyclone.rst, external/scanpy.external.tl.palantir.rst, external/scanpy.external.tl.phate.rst, external/scanpy.external.tl.phenograph.rst, external/scanpy.external.tl.sandbag.rst, index.rst, installation.rst, references.rst, release_notes.rst, tutorials.rst
#   [autosummary] generating autosummary for: /Users/isaac/github/scanpy/docs/api/scanpy.metrics.confusion_matrix.rst, /Users/isaac/github/scanpy/docs/api/scanpy.metrics.gearys_c.rst
#   building [mo]: targets for 0 po files that are out of date
#   building [html]: targets for 6 source files that are out of date
#   updating environment:
#   [extensions changed ('5')]
#   242 added, 0 changed, 0 removed
#   reading sources... [  0%] api/index
# Loaded extensions:
#   sphinx.ext.mathjax (2.2.1) from /usr/local/lib/python3.7/site-packages/sphinx/ext/mathjax.py
#   sphinxcontrib.applehelp (1.0.1) from /usr/local/lib/python3.7/site-packages/sphinxcontrib/applehelp/__init__.py
#   sphinxcontrib.devhelp (1.0.1) from /usr/local/lib/python3.7/site-packages/sphinxcontrib/devhelp/__init__.py
#   sphinxcontrib.htmlhelp (1.0.2) from /usr/local/lib/python3.7/site-packages/sphinxcontrib/htmlhelp/__init__.py
#   sphinxcontrib.serializinghtml (1.1.3) from /usr/local/lib/python3.7/site-packages/sphinxcontrib/serializinghtml/__init__.py
#   sphinxcontrib.qthelp (1.0.2) from /usr/local/lib/python3.7/site-packages/sphinxcontrib/qthelp/__init__.py
#   alabaster (0.7.12) from /usr/local/lib/python3.7/site-packages/alabaster/__init__.py
#   sphinx.ext.autodoc (2.2.1) from /usr/local/lib/python3.7/site-packages/sphinx/ext/autodoc/__init__.py
#   sphinx.ext.intersphinx (2.2.1) from /usr/local/lib/python3.7/site-packages/sphinx/ext/intersphinx.py
#   sphinx.ext.doctest (2.2.1) from /usr/local/lib/python3.7/site-packages/sphinx/ext/doctest.py
#   sphinx.ext.coverage (2.2.1) from /usr/local/lib/python3.7/site-packages/sphinx/ext/coverage.py
#   sphinx.ext.napoleon (2.2.1) from /usr/local/lib/python3.7/site-packages/sphinx/ext/napoleon/__init__.py
#   sphinx.ext.autosummary (2.2.1) from /usr/local/lib/python3.7/site-packages/sphinx/ext/autosummary/__init__.py
#   sphinx_autodoc_typehints (unknown version) from /usr/local/lib/python3.7/site-packages/sphinx_autodoc_typehints.py
#   scanpydoc.autosummary_generate_imported (0.4.4) from /usr/local/lib/python3.7/site-packages/scanpydoc/autosummary_generate_imported.py
#   scanpydoc.definition_list_typed_field (0.4.4) from /usr/local/lib/python3.7/site-packages/scanpydoc/definition_list_typed_field.py
#   scanpydoc.elegant_typehints (0.4.4) from /usr/local/lib/python3.7/site-packages/scanpydoc/elegant_typehints/__init__.py
#   scanpydoc.rtd_github_links (0.4.4) from /usr/local/lib/python3.7/site-packages/scanpydoc/rtd_github_links.py
#   scanpydoc (0.4.4) from /usr/local/lib/python3.7/site-packages/scanpydoc/__init__.py
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/sphinx/cmd/build.py", line 276, in build_main
    app.build(args.force_all, filenames)
  File "/usr/local/lib/python3.7/site-packages/sphinx/application.py", line 346, in build
    self.builder.build_update()
  File "/usr/local/lib/python3.7/site-packages/sphinx/builders/__init__.py", line 299, in build_update
    len(to_build))
  File "/usr/local/lib/python3.7/site-packages/sphinx/builders/__init__.py", line 311, in build
    updated_docnames = set(self.read())
  File "/usr/local/lib/python3.7/site-packages/sphinx/builders/__init__.py", line 418, in read
    self._read_serial(docnames)
  File "/usr/local/lib/python3.7/site-packages/sphinx/builders/__init__.py", line 439, in _read_serial
    self.read_doc(docname)
  File "/usr/local/lib/python3.7/site-packages/sphinx/builders/__init__.py", line 479, in read_doc
    doctree = read_doc(self.app, self.env, self.env.doc2path(docname))
  File "/usr/local/lib/python3.7/site-packages/sphinx/io.py", line 326, in read_doc
    pub.publish()
  File "/usr/local/lib/python3.7/site-packages/docutils/core.py", line 217, in publish
    self.settings)
  File "/usr/local/lib/python3.7/site-packages/sphinx/io.py", line 114, in read
    self.parse()
  File "/usr/local/lib/python3.7/site-packages/docutils/readers/__init__.py", line 77, in parse
    self.parser.parse(self.input, document)
  File "/usr/local/lib/python3.7/site-packages/sphinx/parsers.py", line 95, in parse
    self.statemachine.run(inputlines, document, inliner=self.inliner)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 171, in run
    input_source=document['source'])
  File "/usr/local/lib/python3.7/site-packages/docutils/statemachine.py", line 239, in run
    context, state, transitions)
  File "/usr/local/lib/python3.7/site-packages/docutils/statemachine.py", line 460, in check_line
    return method(match, context, next_state)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 2771, in underline
    self.section(title, source, style, lineno - 1, messages)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 327, in section
    self.new_subsection(title, lineno, messages)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 395, in new_subsection
    node=section_node, match_titles=True)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 282, in nested_parse
    node=node, match_titles=match_titles)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 196, in run
    results = StateMachineWS.run(self, input_lines, input_offset)
  File "/usr/local/lib/python3.7/site-packages/docutils/statemachine.py", line 239, in run
    context, state, transitions)
  File "/usr/local/lib/python3.7/site-packages/docutils/statemachine.py", line 460, in check_line
    return method(match, context, next_state)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 2771, in underline
    self.section(title, source, style, lineno - 1, messages)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 327, in section
    self.new_subsection(title, lineno, messages)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 395, in new_subsection
    node=section_node, match_titles=True)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 282, in nested_parse
    node=node, match_titles=match_titles)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 196, in run
    results = StateMachineWS.run(self, input_lines, input_offset)
  File "/usr/local/lib/python3.7/site-packages/docutils/statemachine.py", line 239, in run
    context, state, transitions)
  File "/usr/local/lib/python3.7/site-packages/docutils/statemachine.py", line 460, in check_line
    return method(match, context, next_state)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 2344, in explicit_markup
    nodelist, blank_finish = self.explicit_construct(match)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 2356, in explicit_construct
    return method(self, expmatch)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 2099, in directive
    directive_class, match, type_name, option_presets)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 2148, in run_directive
    result = directive_instance.run()
  File "/usr/local/lib/python3.7/site-packages/sphinx/ext/autosummary/__init__.py", line 240, in run
    items = self.get_items(names)
  File "/usr/local/lib/python3.7/site-packages/sphinx/ext/autosummary/__init__.py", line 349, in get_items
    documenter.add_content(None)
  File "/usr/local/lib/python3.7/site-packages/sphinx/ext/autodoc/__init__.py", line 471, in add_content
    for i, line in enumerate(self.process_doc(docstrings)):
  File "/usr/local/lib/python3.7/site-packages/sphinx/ext/autodoc/__init__.py", line 441, in process_doc
    self.options, docstringlines)
  File "/usr/local/lib/python3.7/site-packages/sphinx/application.py", line 450, in emit
    return self.events.emit(event, *args)
  File "/usr/local/lib/python3.7/site-packages/sphinx/events.py", line 103, in emit
    results.append(callback(self.app, *args))
  File "/usr/local/lib/python3.7/site-packages/scanpydoc/elegant_typehints/return_tuple.py", line 46, in process_docstring
    ret_types = get_tuple_annot(get_type_hints(obj).get("return"))
  File "/usr/local/Cellar/python/3.7.5/Frameworks/Python.framework/Versions/3.7/lib/python3.7/typing.py", line 996, in get_type_hints
    'or function.'.format(obj))
TypeError: <dispatched gearys_c> is not a module, class, method, or function.

```

</details>

Reply:
I can’t really se the signature because of the corrupted ANSI escape sequences. can you copy it from a terminal that lets you copy plain text?

the fix in scanpydoc seems easy!

Reply:
I updated the header for the docstrings in the examples gave.

Reply:
OK, that looks good! Can you test it with theislab/scanpydoc#12?

Reply:
I've responded on that PR